File: binary-search/matrix-median.cpp
int Solution::findMedian(vector<vector<int> > &A) {
    int n = A.size(), m = A[0].size();
    int count = (n * m + 1) / 2;

    int min = A[0][0], max = A[0][m-1];
    for(int i = 1; i < n; i++){
        if(A[i][0] < min) min = A[i][0];
        if(A[i][m-1] > max) max = A[i][m-1];
    }

    while(min < max){
        int mid = (min + max) / 2;
        int k = 0;

        for(int i = 0; i < n; ++i){
           k += upper_bound(A[i].begin(), A[i].end(), mid) - A[i].begin(); 
        }

        if(k < count) min = mid + 1;
        else max = mid;
    }

    return min;
}

File: binary-search/matrix-median.py
class Solution:
    def binary_search(self, A, key):
        low = 0
        high = len(A) - 1

        while low <= high:
            mid = low + (high - low) // 2
            if A[mid] <= key: low = mid + 1
            else: high = mid - 1

        return low

    def findMedian(self, A):
        n = len(A)
        m = len(A[0])
        median_count = (n * m + 1) // 2

        low = 10 ** 9
        high = 0
        for i in A:
            low = min(low, i[0])
            high = max(high, i[m-1])

        while low < high:
            mid = low + (high - low) // 2
            count = 0

            for i in A:
                count += self.binary_search(i, mid)

            if count < median_count: low = mid + 1
            else: high = mid

        return low



